Masked modeling

jmeld opened the pull request:
1. in config/training_args/base.yaml changed "evaluation_steps" to "eval_steps" to be comply with warning about soon to be deprecated key word arguments.
2. created src/model/custom_electra which holds src/model/custom_electra/electra.py and src/model/custom_electra/__init__.py. These allow for passing our custom electra model training.

shanest commented on src/model/custom_electra/electra.py:
```diff
@@ -0,0 +1,114 @@
+        #super().__init__(pretrained_model_name_or_path)
+        super().__init__(**kwargs)
+        self.vocab_size= vocab_size
+        self.generator_config= generator_model
```
Should this be `self.generator_model` instead of `.generator_config` (and same for the discriminator below?

shanest replied:
Actually, I think the variables in the `kwargs` should just be named `_config` instead of the `_model` as they are now

shanest commented on src/model/custom_electra/electra.py:
```diff
@@ -0,0 +1,114 @@
+
+    def __init__(
+        self,
+        generator_model: transformers.DebertaV2Config,
```
(1) rename to `generator_config` (and same for `discriminator`) below.  (2) Just to be fully clear, we don't want to just replace `ElectraX` with `DebertaX`, but rather the most general possible thing that consumes both.  For configs, I honestly think that's just `PretrainedConfig` here

shanest commented on src/model/custom_electra/electra.py:
```diff
@@ -0,0 +1,114 @@
+        # or pass the generator/discriminators themselves
+        #base_gen_config = ElectraConfig.from_pretrained('google/electra-base-generator')
+        base_gen_config= AutoConfig.from_pretrained(config.generator_config)
+        self.generator: ElectraForMaskedLM = AutoModelForMaskedLM.from_config(base_gen_config)
```
Type annotation here should just be either `: PretrainedModel` or `AutoModelForMaskedLM`

shanest commented on src/model/custom_electra/electra.py:
```diff
@@ -0,0 +1,114 @@
+
+        #base_disc_config = ElectraConfig.from_pretrained('google/electra-base-discriminator')
+        base_disc_config = AutoConfig.from_pretrained(config.discriminator_config)
+        self.discriminator: ElectraForPreTraining = AutoModelForMaskedLM.from_config(base_disc_config)
```
For the discriminator, we don't want `MaskedLM`, since the task here is replaced token detection.  I'd say there are two options: (i) use `AutoModelForTokenClassification` instead of `AutoModelForMaskedLM`, or (ii) just use `AutoModel`, and add the binary token classification head to this class in `__init__` and `forward`.  Option (i) is probably the smoothest

shanest commented on src/model/custom_electra/electra.py:
```diff
@@ -0,0 +1,114 @@
+                input_ids: Optional[torch.Tensor] = None, 
+                attention_mask: Optional[torch.Tensor] = None, 
+                token_type_ids: Optional[torch.Tensor] = None, 
+                labels: Optional[torch.Tensor] = None) -> ElectraForPreTrainingOutput:
```
As per above: return type could be `TokenClassifierOutput`: https://huggingface.co/docs/transformers/v4.44.2/en/main_classes/output#transformers.modeling_outputs.TokenClassifierOutput

shanest commented on src/model/custom_electra/electra.py:
```diff
@@ -0,0 +1,114 @@
+        # loss
+        loss = generator_output.loss + discriminator_output.loss * self.lambda_loss #might have to rewrite as loss= generator_output.loss.loss + discriminator_output.loss.loss * self.lambda_loss or it might be that discriminator_output.loss.loss is the only one that needs to be .loss twice
+        # must return some output.loss in order to work with trainer
+        return ElectraForPreTrainingOutput(
```
As above, I think this can just be a `TokenClassifierOutput` for max generality https://huggingface.co/docs/transformers/v4.44.2/en/main_classes/output#transformers.modeling_outputs.TokenClassifierOutput

shanest reviewed (changes requested):
Looking great!  Just a few little changes here and there to get the right level of generality still

jmeld commented:
Thank you for the prompt and informative suggestions.

shanest reviewed (changes requested):
Looks great!  Can you run some quick tests of it first, e.g. passing some inputs through?  I think there should be some errors around `TokenClassifierOutput` and the like, since those aren't imported up top, so doing that kind of running as a test should turn up all of them.  (Running a linter like `mypy` or `pyright` will also turn these up)

jmeld commented:
Thank you! I implemented the suggestions, and the code runs. I'm not sure why, but pyright highlights some code as breaking type constraints, but I think it might be mistaken.

shanest commented:
Hi Jon: just circling back to what we discussed in the meeting the other day.  Can you and/or Diana work on integrating the custom Electra/RTD classes in BabyLM repo and getting things working there?  I'd like to merge this PR into main, but without the custom model classes :)

jmeld commented:
Yes. I have

1. Removed the src/model/custom_electra/__init__.py and src/model/custom_electra/electra.py files from this lm-training repo to clean it up.
2. In the babylm branch da/electra I added a new directory babylm and made it into a package by adding babylm/__init__.py, and added a sub-package named model with babylm/model/electra.py and babylm/model/__init__.py

kharaldsson commented:
> Yes. I have
> 
> 1. Removed the src/model/custom_electra/**init**.py and src/model/custom_electra/electra.py files from this lm-training repo to clean it up.
> 2. Made the changes you suggest here to electra.py in the babylm git repo's da/electra branch. So this pull request should clean up the lm-training repo
> 3. Added some new tokenizer training configs for training masked language modeling but adding a MASK, CLS, and SEP token instead of just a MASK token (as the previous masked training model does).

@shanest are there any outstanding comments on this PR? I'm assuming it's a blocker for bablym?

shanest commented:
Thanks for digging this back up!  And thanks Jon for cleaning things up. Yet, looks good to me now, will merge :).  We'll need to go back to the babylm repo and check out the modeling PR over there for the Next Token Prediction stuff
